### PR TITLE
[BEAM-6527] Use Gradle to parallel Python tox tests

### PR DESCRIPTION
**Pull request by markflyhigh**
This PR is to roll-forward https://github.com/apache/beam/pull/7675 with fixes suggested by @kkucharc:
- Avoid recursive copy of non-source files under `sdks/python/apache_beam` in `BeamModulePlugin.groovy` [line 1619](https://github.com/apache/beam/pull/8067/files#diff-23833058cbf2c1172b90e7764032aa59R1619).
- `sdist` task failed to collect singleFile due to existence of `srcs` under build directory. Fix at `BeamModulePlugin.groovy` [line 1642](https://github.com/apache/beam/pull/8067/files#diff-23833058cbf2c1172b90e7764032aa59R1642). Error looks like:
```
Caused by: java.lang.IllegalStateException: Expected directory '/Users/kasia/Repos/beam/sdks/python/build' to contain exactly one file, however, it contains more than one file.
        at org.gradle.api.internal.file.AbstractFileCollection.getSingleFile(AbstractFileCollection.java:65)
        at org.apache.beam.gradle.BeamModulePlugin$_apply_closure20$_closure136$_closure145.doCall(BeamModulePlugin.groovy:1644)
```

+R: @aaltay 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by markflyhigh** [comment]
Run Gradle Publish

**Comment by markflyhigh** [comment]
This PR is ready for review.

`./gradlew publish`(nightly snapshot job) is failed due to reach to Jenkins job timeout (140 mins). No error is fund in Python tests.

**Comment by markflyhigh** [comment]
+cc: @kkucharc @ajamato @tvalentyn

**Comment by markflyhigh** [comment]
Run Gradle Publish

**Comment by tvalentyn** [comment]
Run Gradle Publish

**Comment by tvalentyn** [comment]
LGTM, not sure why gradle publish failed. It would be easier to review if you separated new changes and   previously reviewed commit into separate commit, but looks like filepaths are the only things that changed.

**Comment by tvalentyn** [comment]
Run Gradle Publish

**Comment by adude3141** [comment]
>  not sure why gradle publish failed.

As @markflyhigh already stated, the publish task is killed by Jenkins after 140 min (2h20min). Not sure what did cause the build time increase here. Probably has nothing to do with this commit, but was hidden before as build was already failing on earlier tasks.

So we probably need to increase that timeout, at least temporarily, until we are able to improve test performance.

**Comment by aaltay** [comment]
Is the test error a flake? Can we rerun it?

**Comment by adude3141** [comment]
I do not believe its a flake. The whole build just takes more than 140mins on Jenkins right now.

This might be caused by some added tests. The last successful [1] publish on Feb, 24 took 132mins already, so it seems we need to increase the timeout here.

@markflyhigh Could you increase this timeout with this PR, as - at least to my understanding - the build fails on master till we get this PR merged 

I have no idea how much is required, but I d go for 3 (or even 4?) hours.

[1] https://builds.apache.org/job/beam_Release_NightlySnapshot/344/

**Comment by markflyhigh** [comment]
Yes, it doesn't seem to be a flake. I'll increase the timeout to 200mins in this PR to see if the job can pass successfully. I feel 3 - 4 hours is acceptable currently since it only runs once a day.

**Comment by markflyhigh** [comment]
Run Gradle Publish

**Comment by markflyhigh** [comment]
Run Seed Job

**Comment by markflyhigh** [comment]
Run Gradle Publish

**Comment by markflyhigh** [comment]
Run Python PreCommit

**Comment by markflyhigh** [comment]
Run Seed Job

**Comment by markflyhigh** [comment]
Run Gradle Publish

**Comment by markflyhigh** [comment]
Run Java PreCommit

**Comment by markflyhigh** [comment]
"./gradlew publish" failed since python36 and 37 are missing on Beam12. This is a infra issue and I start [a discussion] on dev@(https://lists.apache.org/thread.html/529a3626eb9e8f5795b271b9d06ba34242fcaf771fdc8f596eaec229@%3Cdev.beam.apache.org%3E). Gradle scan shows that python load test was built successfully.

PTAL @aaltay @tvalentyn @adude3141

**Comment by adude3141** [comment]
Run Gradle Publish

**Comment by adude3141** [comment]
Run Gradle Publish

**Comment by markflyhigh** [comment]
This time nightly snapshot failed ([job link](https://scans.gradle.com/s/3dm6dt5ut2aee)) on `publishMavenJavaPublicationToMavenRepository` in many java projects. According to the build history of this job, it's been failed/broken since Feb.

**Comment by markflyhigh** [comment]
Run Python Load Tests Smoke

**Comment by tvalentyn** [comment]
LGTM

**Comment by markflyhigh** [comment]
Run Python PostCommit

**Comment by markflyhigh** [comment]
I think current tests are good enough for the coverage: 

- precommit
- postcommit2/3
- load tests
- Portable_Python
- Python_PVR_Flink

Can we merge it? @aaltay @adude3141

**Comment by adude3141** [comment]
Thanks, @markflyhigh, I have to admit, you put lot of effort into testing this. And it is most likely, that no other will pass the publish.

But as it improves the overall performance, lets merge and fix outstanding issues separately.

Again, thanks a lot for the effort you put into this.

**Comment by adude3141** [comment]
@markflyhigh
While running build on my local machine, I've seen tons of additional log output cluttering console.

Is there any option to quieten that? It is difficult to follow and I prefer to add that '--info' resp '--debug' flags if required.

**Comment by markflyhigh** [comment]
Do you want to quite/control logs that are generated from python commands? According to [Gradle doc](https://docs.gradle.org/current/userguide/logging.html#sec:sending_your_own_log_messages): `Gradle redirects anything written to standard output to its logging system at the QUIET log level.` Seems there is no [option](https://docs.gradle.org/current/userguide/logging.html#sec:choosing_a_log_level) to disable this level.

I think you may need to do it inside python and then integrate with Gradle if possible.

**Comment by adude3141** [comment]
Comparing e.g. java precommit build [1] with python [2] I was under the impression, that the log output heavily increased with this commit.

Just wondering, if you got the same feeling. From my point of view, most output is not helpful, e.g. all this 'creating...', 'copying', 'adding' just clutters console. 

You are right regarding gradles default behaviour. But we might consider adapting that to something more sensible on task level? Either by quietening the tools delegated to - which might have some drawbacks - or we could add a
` logging.captureStandardOutput org.gradle.api.logging.LogLevel.INFO` 
(or even better DEBUG?) to the 'sdist' task which will quieten things a bit. And similar stuff with other tasks.

Did not look deeper into that, though. And of course, it has a lot to do with daily dev - which I am not doing on python currently -, whether that output should be seen or could be considered clutter (which might be enabled if required by adding '-I' or '-q' param.

[1] https://builds.apache.org/job/beam_PreCommit_Java_Phrase/827/consoleFull
[2] https://builds.apache.org/job/beam_PreCommit_Python_Commit/5194/consoleFull

File: sdks/python/test-suites/tox/py36/build.gradle [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+/**
+ * Unit tests for Python 3.6
+ */
+
+plugins { id 'org.apache.beam.module' }
+applyPythonNature()
+
+toxTask "testPython36", "py36"
```
Comment (udim) [inline sdks/python/test-suites/tox/py36/build.gradle]: How come there's no py36-gcp?

Comment (tvalentyn) [reply sdks/python/test-suites/tox/py36/build.gradle]: py36 cython and gcp suites are being worked on and will be added.

Comment (tvalentyn) [reply sdks/python/test-suites/tox/py36/build.gradle]: Specifically, in https://github.com/apache/beam/pull/8036

Comment (markflyhigh) [reply sdks/python/test-suites/tox/py36/build.gradle]: Note, this PR is target on reorganizing tasks to make parallel work, not enable new test suite, which requires more manual effort to investigate failure and do the fix. @RobbeSneyders has more context on new suite work.

**Pull request merged**